File: 3606-coupon-code-validator.py
class Solution:
    def validateCoupons(self, code: list[str], businessLine: list[str], isActive: list[bool]) -> list[str]:
        '''
        We create 4 lists, one for each business line, containing the valid codes for each business line
        Next, we go through each code, check if code is valid and is active.
            - if so, we append the code to the list for the business line
            
        At the end, we return a list of sorted codes for each business line
        '''
        electronics, grocery, pharmacy, restaurant = [], [], [], []
        for i, c in enumerate(code):
            if not c or not c.replace('_', 'a').isalnum() or not isActive[i]:
                continue
            business = businessLine[i]
            if business == 'electronics':
                electronics.append(c)
            elif business == 'grocery':
                grocery.append(c)
            elif business == 'pharmacy':
                pharmacy.append(c)
            elif business == 'restaurant':
                restaurant.append(c)
        
        return sorted(electronics) + sorted(grocery) + sorted(pharmacy) + sorted(restaurant)
        
sol = Solution()
print(f'output: {sol.validateCoupons(code = ["SAVE20","","PHARMA5","SAVE@20"], businessLine = ["restaurant","grocery","pharmacy","restaurant"], isActive = [True,True,True,True])}, expected: ["PHARMA5","SAVE20"]')
print(f'output: {sol.validateCoupons(code = ["GROCERY15","ELECTRONICS_50","DISCOUNT10"], businessLine = ["grocery","electronics","invalid"], isActive = [False,True,True])}, expected: ["ELECTRONICS_50"]')

File: 3606-coupon-code-validator.java
import java.util.*;

class CouponCodeValidator {
    public List<String> validateCoupons(String[] code, String[] businessLine, boolean[] isActive) {

        List<String> electronics = new ArrayList<>();
        List<String> grocery = new ArrayList<>();
        List<String> pharmacy = new ArrayList<>();
        List<String> restaurant = new ArrayList<>();

        for (int i = 0; i < code.length; i++) {
            String c = code[i];
            if (c.isBlank() || !c.matches("[A-Za-z0-9_]+") || !isActive[i]) {
                continue;
            }
            String business = businessLine[i];
            switch (business) {
                case "electronics":
                    electronics.add(c);
                    break;
                case "grocery":
                    grocery.add(c);
                    break;
                case "pharmacy":
                    pharmacy.add(c);
                    break;
                case "restaurant":
                    restaurant.add(c);
                    break;
                default:
                    break;
            }
        }
        Comparator<String> comparator = new Comparator<>() {
            @Override
            public int compare(String o1, String o2) {
                return o1.compareTo(o2);
            }
            
        };
        List<String> res = new ArrayList<>();
        electronics.sort(comparator);
        grocery.sort(comparator);
        pharmacy.sort(comparator);
        restaurant.sort(comparator);
        res.addAll(electronics);
        res.addAll(grocery);
        res.addAll(pharmacy);
        res.addAll(restaurant);
        return res;
    }

    public static void main(String[] args) {
        CouponCodeValidator sol = new CouponCodeValidator();
        System.out.println("Output: " + sol.validateCoupons(new String[] { "SAVE20", "", "PHARMA5", "SAVE@20" },
                new String[] { "restaurant", "grocery", "pharmacy", "restaurant" },
                new boolean[] { true, true, true, true }).toString() + ", expected: ['PHARMA5','SAVE20']");
        System.out.println("Output: "
                + sol.validateCoupons(new String[] { "GROCERY15", "ELECTRONICS_50", "DISCOUNT10" },
                        new String[] { "grocery", "electronics", "invalid" }, new boolean[] { false, true, true })
                + ", expected: ['ELECTRONICS_50']");
    }
}
